pattern/constraint/thing/RelationConstraint.java

```diff
@@ -136,19 +137,24 @@ public int hashCode() {
     }
 
     @Override
-    public AlphaEquivalence alphaEquals(RelationConstraint that) {
-        return AlphaEquivalence.valid()
-                .validIf(players().size() == that.players().size())
-                .addOrInvalidate(() -> Iterators.permutation(players()).stream().map(playersPermutation -> {
-                    Iterator<RolePlayer> thisRolePlayersIt = playersPermutation.iterator();
-                    Iterator<RolePlayer> thatRolePlayersIt = that.players().iterator();
-                    AlphaEquivalence permutationMap = AlphaEquivalence.valid();
-                    while (thisRolePlayersIt.hasNext() && thatRolePlayersIt.hasNext()) {
-                        permutationMap = permutationMap.validIfAlphaEqual(thisRolePlayersIt.next(), thatRolePlayersIt.next());
-                        if (!permutationMap.isValid()) return permutationMap;
-                    }
-                    return permutationMap;
-                }).filter(AlphaEquivalence::isValid).findFirst().orElse(AlphaEquivalence.invalid()));
+    public FunctionalIterator<AlphaEquivalence> alphaEquals(RelationConstraint that) {
+        return AlphaEquivalence.empty()
+                .alphaEqualIf(players().size() == that.players().size())
+                .flatMap(a -> roleplayerEquivalences(that).flatMap(a::extendIfCompatible));
+    }
+
+    private FunctionalIterator<AlphaEquivalence> roleplayerEquivalences(RelationConstraint that) {
+        return Iterators.permutation(players()).flatMap(playersPermutation -> {
+            Iterator<RolePlayer> thisRolePlayersIt = playersPermutation.iterator();
+            Iterator<RolePlayer> thatRolePlayersIt = that.players().iterator();
```

Comment:
not our style to us `It` to denote the type, since we generally work in an IDE that can tell us the type if we have forgotten it :)